cli/context/docker/load.go

```diff
@@ -0,0 +1,192 @@
+			}
+			switch f {
+			case caKey:
+				tlsData.CA = data
+			case certKey:
+				tlsData.Cert = data
+			case keyKey:
+				tlsData.Key = data
+			default:
+				logrus.Warnf("unknown file %s in context %s tls bundle", f, c.ContextName)
+			}
+		}
+		return Endpoint{EndpointMeta: *c, TLSData: &tlsData}, nil
+	}
+	return Endpoint{EndpointMeta: *c}, nil
+}
+
+// tlsConfig extracts a context docker endpoint TLS config
+func (c *Endpoint) tlsConfig() (*tls.Config, error) {
+	if c.TLSData == nil && !c.SkipTLSVerify {
+		// there is no specific tls config
+		return nil, nil
+	}
+	var tlsOpts []func(*tls.Config)
+	if c.TLSData != nil && c.TLSData.CA != nil {
+		certPool := x509.NewCertPool()
+		if !certPool.AppendCertsFromPEM(c.TLSData.CA) {
+			return nil, errors.New("failed to retrieve context tls info: ca.pem seems invalid")
+		}
+		tlsOpts = append(tlsOpts, func(cfg *tls.Config) {
+			cfg.RootCAs = certPool
+		})
+	}
+	if c.TLSData != nil && c.TLSData.Key != nil && c.TLSData.Cert != nil {
+		keyBytes := c.TLSData.Key
+		pemBlock, _ := pem.Decode(keyBytes)
+		if pemBlock == nil {
+			return nil, fmt.Errorf("no valid private key found")
+		}
+
+		var err error
+		if x509.IsEncryptedPEMBlock(pemBlock) {
+			keyBytes, err = x509.DecryptPEMBlock(pemBlock, []byte(c.TLSPassword))
+			if err != nil {
+				return nil, errors.Wrap(err, "private key is encrypted, but could not decrypt it")
+			}
+			keyBytes = pem.EncodeToMemory(&pem.Block{Type: pemBlock.Type, Bytes: keyBytes})
+		}
+
+		x509cert, err := tls.X509KeyPair(c.TLSData.Cert, keyBytes)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to retrieve context tls info")
+		}
+		tlsOpts = append(tlsOpts, func(cfg *tls.Config) {
+			cfg.Certificates = []tls.Certificate{x509cert}
+		})
+	}
+	if c.SkipTLSVerify {
+		tlsOpts = append(tlsOpts, func(cfg *tls.Config) {
+			cfg.InsecureSkipVerify = true
+		})
+	}
+	return tlsconfig.ClientDefault(tlsOpts...), nil
+}
+
+//ConfigureClient configures a docker client
+func (c *Endpoint) ConfigureClient(cli *client.Client) error {
+	if c.Host != "" {
+		helper, err := connhelper.GetConnectionHelper(c.Host)
+		if err != nil {
+			return err
+		}
+		if helper == nil {
+			if err := client.WithHost(c.Host)(cli); err != nil {
+				return err
+			}
+		} else {
+			httpClient := &http.Client{
+				// No tls
+				// No proxy
+				Transport: &http.Transport{
+					DialContext: helper.Dialer,
+				},
+			}
+			if err := client.WithHTTPClient(httpClient)(cli); err != nil {
+				return err
+			}
+			if err := client.WithHost(helper.Host)(cli); err != nil {
+				return err
+			}
+			if err := client.WithDialContext(helper.Dialer)(cli); err != nil {
+				return err
+			}
+		}
+	}
+	tlsConfig, err := c.tlsConfig()
+	if err != nil {
+		return err
+	}
+	if tlsConfig != nil {
+		httpClient := cli.HTTPClient()
+		if transport, ok := httpClient.Transport.(*http.Transport); ok {
+			transport.TLSClientConfig = tlsConfig
```

Comment:
Even if I don't know how to refactor that, it feels really smelly...

Reply:
yes, I know :(. impossible without modifying moby/moby